File: app/src/main/java/com/codestallions/spacemmo/ui/activities/BaseActivity.java
package com.codestallions.spacemmo.ui.activities;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.codestallions.spacemmo.SessionManager;
import com.codestallions.spacemmo.SpaceMMO;

public abstract class BaseActivity extends AppCompatActivity {

    private static final long SESSION_CHECK_INTERVAL = 30000;
    private Handler sessionCheckHandler;

    private Runnable sessionCheckRunnable = new Runnable()
    {
        public void run()
        {
            if (SpaceMMO.getSession().isLoginExpired(getApplicationContext()) &&
                    SpaceMMO.getAuth().getCurrentUser() != null) {
                SpaceMMO.getAuth().signOut();
                navigateToLogin();
            }
            sessionCheckHandler.postDelayed(this, SESSION_CHECK_INTERVAL);
        }
    };

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        sessionCheckHandler = new Handler(Looper.getMainLooper());
    }

    @Override
    protected void onStart() {
        super.onStart();
    }

    @Override
    protected void onResume() {
        super.onResume();
        startSessionWatcher();
    }

    private void startSessionWatcher() {
        sessionCheckHandler.postDelayed(sessionCheckRunnable, SESSION_CHECK_INTERVAL);
    }

    public void navigateToLogin() {
        Intent intent = new Intent(this, LoginActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        startActivity(intent);
    }
}

File: app/src/main/java/com/codestallions/spacemmo/ui/activities/BaseActivity.kt
package com.codestallions.spacemmo.ui.activities

import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import androidx.appcompat.app.AppCompatActivity
import com.codestallions.spacemmo.SpaceMMO
import com.codestallions.spacemmo.ui.activities.LoginActivity

abstract class BaseActivity : AppCompatActivity() {
    private var sessionCheckHandler: Handler? = null
    private val sessionCheckRunnable: Runnable = object : Runnable {
        override fun run() {
            if (SpaceMMO.getSession().isLoginExpired(applicationContext) &&
                    SpaceMMO.getAuth().currentUser != null) {
                SpaceMMO.getAuth().signOut()
                navigateToLogin()
            }
            sessionCheckHandler!!.postDelayed(this, SESSION_CHECK_INTERVAL)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        sessionCheckHandler = Handler(Looper.getMainLooper())
    }

    override fun onResume() {
        super.onResume()
        startSessionWatcher()
    }

    private fun startSessionWatcher() {
        sessionCheckHandler!!.postDelayed(sessionCheckRunnable, SESSION_CHECK_INTERVAL)
    }

    fun navigateToLogin() {
        val intent = Intent(this, LoginActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
        startActivity(intent)
    }

    companion object {
        private const val SESSION_CHECK_INTERVAL: Long = 30000
    }
}